module/os/linux/zfs/zpl_super.c

```diff
@@ -384,7 +384,18 @@ zpl_prune_sb(uint64_t nr_to_scan, void *arg)
 	struct super_block *sb = (struct super_block *)arg;
 	int objects = 0;
 
-	(void) -zfs_prune(sb, nr_to_scan, &objects);
+	/*
+	 * deactivate_locked_super calls shrinker_free and only then
+	 * sops->kill_sb cb, resulting in UAF on umount when trying to reach
+	 * for the shrinker functions in zpl_prune_sb of in-umount dataset.
+	 * Increment if s_active is not zero, but don't prune if it is -
+	 * umount could be underway.
+	 */
+	if (atomic_inc_not_zero(&sb->s_active)) {
```

Comment:
`zfs_preumount` gets called in `zpl_kill_sb`, which itself gets called after the kernel has done `shrinker_free`, so that'd still leave +- similar window for NPE; we'd need something that'd get called before the `.kill_sb`, but I didn't find anything, so I went looking for how other FS do it and `s_active` seems to be it